sacred/observers/s3_observer.py

```diff
@@ -0,0 +1,328 @@
+#!/usr/bin/env python
+# coding=utf-8
+
+import json
+import os
+import os.path
+
+import boto3
+from botocore.errorfactory import ClientError
+
+from sacred.commandline_options import CommandLineOption
+from sacred.dependencies import get_digest
+from sacred.observers.base import RunObserver
+from sacred.serializer import flatten
+import re
+import socket
+
+DEFAULT_S3_PRIORITY = 20
+
+
+def _is_valid_bucket(bucket_name):
+    # See https://docs.aws.amazon.com/awscloudtrail/latest/userguide/
+    # cloudtrail-s3-bucket-naming-requirements.html
+    if len(bucket_name) < 3 or len(bucket_name) > 63:
+        return False
+    if '..' in bucket_name or '.-' in bucket_name or '-.' in bucket_name:
```

Comment:
```suggestion
   forbidden_words = {"..", ".-", "-.", "_"}
   if any(word in bucket_name for word in forbidden_words):
```
I believe the "cannot contain underscores" requirement was missing.

Reply:
I think that the "cannot contain underscores" requirement was handled by there being an explicit whitelist of allowed character types (lowercase characters, digits, dashes), where underscores don't fall into any of those whitelisted categories

Reply:
That said, I think the old code was incorrect in that it didn't have periods as one of the whitelisted characters. I've refactored this to have the notion of "labels", which are distinct alphanumeric chunks separated by periods, and I think this makes things cleaner because it matches the abstraction that AWS itself uses